### PR TITLE
Fixed #26821 -- use correct strip functionality for EmailField and URLField

### PR DESCRIPTION
**Pull request by PriyWerry**
https://code.djangoproject.com/ticket/26821

### CODE REVIEW

File: django/forms/fields.py [hunk 1]
```diff
@@ -532,8 +534,18 @@ class EmailField(CharField):
     widget = EmailInput
     default_validators = [validators.validate_email]
 
+    def __init__(self, *args, **kwargs):
+        strip = kwargs.pop('strip', None)
+        if strip is not None:
+            warnings.warn(
+                "Passing `strip` as keyword argument is deprecated. "
```
Comment (MarkusH) [inline django/forms/fields.py]: I think something like this is more what we want:
```
"Passing `strip` as an argument to EmailField is deprecated. EmailField values will always be strip of leading/trailing white spaces."
```

File: docs/ref/forms/fields.txt [hunk 2]
```diff
@@ -585,6 +585,11 @@ For each field, we describe the default widget used if you don't specify
     If provided, these arguments ensure that the string is at most or at least the
     given length.
 
+    .. deprecated:: 1.10
+
+        The `strip` keyword argument is deprecated. An `EmailField` automatically
```
Comment (MarkusH) [inline docs/ref/forms/fields.txt]: s/automatically/always/

File: tests/forms_tests/field_tests/test_urlfield.py [hunk 3]
```diff
@@ -151,3 +151,13 @@ def test_urlfield_not_string(self):
     def test_urlfield_normalization(self):
         f = URLField()
         self.assertEqual(f.clean('http://example.com/     '), 'http://example.com/')
+
+    def test_urlfield_strip_false(self):
+        f = URLField(strip=False)
+        with self.assertRaisesMessage(ValidationError, "'Enter a valid URL.'"):
+            f.clean('http://example.com ')
```
Comment (MarkusH) [inline tests/forms_tests/field_tests/test_urlfield.py]: How is this related to this PR?

Comment (PriyWerry) [inline tests/forms_tests/field_tests/test_urlfield.py]: @MarkusH Before the changes this test would fail. Instead of raising a ValdiationError, an AttributeError ('NoneType' object has no attribute 'strip') would be thrown, as a strip() call was still made (on the None value) even though strip=False was set.

In other words -> this is the regression test.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by PriyWerry** [comment]
It seems the following two tests are failing for some builds, but as far as I know they have nothing to do with my changes:

* [decorators.tests.DecoratorsTest.test_deprecated_allow_lazy](https://github.com/django/django/blob/master/tests/decorators/tests.py#L161) (<class 'django.utils.functional.__proxy__'> != <type 'unicode'>)
* [apps.tests.AppsTests.test_model_clash](https://github.com/django/django/blob/master/tests/apps/tests.py#L214) (RuntimeWarning not raised
)

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
Sorry, my fault. The test issue was the lack of a81fe17b3e5bb96c849c4044d6688c08848dec19.

I asked on the ticket about whether we need a deprecation here.

File: django/forms/fields.py [hunk 4]
```diff
@@ -532,8 +534,18 @@ class EmailField(CharField):
...
+                "EmailField values will always be stripped of leading and trailing whitespace.",
+                RemovedInDjango21Warning, stacklevel=2
+            )
+        super(EmailField, self).__init__(*args, strip=True, **kwargs)
+
     def clean(self, value):
-        value = self.to_python(value).strip()
+        value = self.to_python(value)
```
Comment (jdufresne) [inline django/forms/fields.py]: With this change, can the entire `clean()` method be removed from `EmailField` and `URLField`? The parent implementation already [calls `to_python()`](https://github.com/django/django/blob/415ae960bb9f1bdae798023fdce3247d2c938eec/django/forms/fields.py#L158)?

Comment (PriyWerry) [inline django/forms/fields.py]: Good one, removed them in commit 6f09cc1

File: docs/ref/forms/fields.txt [hunk 5]
```diff
@@ -585,6 +585,11 @@ For each field, we describe the default widget used if you don't specify
     If provided, these arguments ensure that the string is at most or at least the
     given length.
 
+    .. versionadded:: 1.11
```
Comment (timgraham) [inline docs/ref/forms/fields.txt]: I don't think it's important to mention the issue here since the option was never documented. A mention in the 1.11 release notes should be sufficient.

File: tests/forms_tests/field_tests/test_emailfield.py [hunk 6]
```diff
@@ -51,3 +51,13 @@ def test_emailfield_min_max_length(self):
         self.assertEqual('alf@foo.com', f.clean('alf@foo.com'))
         with self.assertRaisesMessage(ValidationError, "'Ensure this value has at most 15 characters (it has 20).'"):
             f.clean('alf123456788@foo.com')
+
+    def test_emailfield_strip_on_none_value(self):
+        """EmailField 'correctly' strips None value instead of raising error (Refs #26821)"""
```
Comment (timgraham) [inline tests/forms_tests/field_tests/test_emailfield.py]: I think the test is simple and straightforward enough that a docstring isn't needed. We only mention ticket numbers for complex issues where the context of ticket might be helpful in understanding what's going on.

**Comment by timgraham** [comment]
You may squash commits when updating. Thanks!

**Comment by timgraham** [comment]
merged in a7b5dfd1703a8fbed70b7aca5e6dda092af51154, thanks!
